### PR TITLE
Atomic exchange

### PR DESCRIPTION
**Pull request by Rombur**

### CODE REVIEW

File: core/src/impl/Kokkos_Atomic_Exchange.hpp [hunk 1]
```diff
@@ -419,6 +420,89 @@ void atomic_assign( volatile T * const dest_v , const T val )
...
+{
+  return atomicExch(static_cast<unsigned long long *>(dest), val);
+}
+
+/** \brief  Atomic exchange for any type with compatible size */
+template <typename T>
+__inline__ __device__ T atomic_exchange(
+    volatile T *const dest,
+    typename Kokkos::Impl::enable_if<sizeof(T) == sizeof(int), const T &>::type
+        val)
```
Comment (dalg24) [inline core/src/impl/Kokkos_Atomic_Exchange.hpp:449]: Use `std::enable_if`

File: core/src/impl/Kokkos_Atomic_Exchange.hpp [hunk 2]
```diff
@@ -419,6 +420,89 @@ void atomic_assign( volatile T * const dest_v , const T val )
...
+    volatile T *const dest,
+    typename Kokkos::Impl::enable_if<
+        sizeof(T) != sizeof(int) && sizeof(T) == sizeof(unsigned long long int),
+        const T &>::type val)
+{
+  typedef unsigned long long int type;
+
+  type tmp = atomicExch(static_cast<type *>(const_cast<T *>(dest)),
+                        *static_cast<type *>(const_cast<T *>(&val)));
+  return *static_cast<T *>(&tmp);
```
Comment (dalg24) [inline core/src/impl/Kokkos_Atomic_Exchange.hpp]:
```suggestion
  return static_cast<T &>(tmp);
```

**Comment by Rombur** [comment]
Done

File: core/src/impl/Kokkos_Atomic_Exchange.hpp [hunk 3]
```diff
@@ -419,6 +420,89 @@ void atomic_assign( volatile T * const dest_v , const T val )
...
+
+/** \brief  Atomic exchange for any type with compatible size */
+template <typename T>
+__inline__ __device__ void
+atomic_assign(volatile T *const dest,
+              typename Kokkos::Impl::enable_if<sizeof(T) == sizeof(int),
+                                               const T &>::type val)
+{
+  atomicExch(static_cast<int *>(const_cast<T *>(dest)),
+             *static_cast<int *>(const_cast<T *>(&val)));
```
Comment (dalg24) [inline core/src/impl/Kokkos_Atomic_Exchange.hpp:478]: If new encounter more of these in the future we should introduce function helpers to perform these cast

**Review by dalg24 — APPROVED** [review]

**Pull request merged**